Added clean exec functionality + ACI implementation - V2

erjadi opened the pull request:
Clean implementation of exec for Virtual Kubelet:

- Reusing standard Kubelet Executor interface (and remotecommand functions)
- Implemented ACI provider

Two TODOs are open:

- How to best pass namespace to provider (currently doing ACI specific way, should be changed)
- Expected streamcount is fixed number, but should be based on exec request to apiserver. However when using exec request, createStreams will wait indefinitely (for +1 stream).

This is a new pull request (invalidating https://github.com/virtual-kubelet/virtual-kubelet/pull/212) based on master after dependencies were fixed. 

People who are just interested in trying it out, use erjadi/virtual-kubelet:cleanexec from dockerhub.

robbiezhang commented on providers/azure/aci.go:
```diff
@@ -329,6 +332,69 @@ func (p *ACIProvider) GetContainerLogs(namespace, podName, containerName string,
+// ExecInContainer executes a command in a container in the pod, copying data
+// between in/out/err and the container's stdin/stdout/stderr.
+func (p *ACIProvider) ExecInContainer(name string, uid types.UID, container string, cmd []string, in io.Reader, out, errstream io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize, timeout time.Duration) error {
+	cg, _, _ := p.aciClient.GetContainerGroup(p.resourceGroup, name)
```
check the error. If it failed, please close the out/err streams.

robbiezhang commented on providers/azure/aci.go:
```diff
@@ -329,6 +332,69 @@ func (p *ACIProvider) GetContainerLogs(namespace, podName, containerName string,
+		terminalSize = <-resize // Receive terminal resize event if resize stream is present
+	}
+
+	xcrsp, _ := p.aciClient.LaunchExec(p.resourceGroup, cg.Name, container, cmd[0], terminalSize)
```
check error.

robbiezhang commented on providers/azure/aci.go:
```diff
@@ -329,6 +332,69 @@ func (p *ACIProvider) GetContainerLogs(namespace, podName, containerName string,
+		go func() {
+			for {
+				var msg = make([]byte, 512)
+				_, err := in.Read(msg)
```
I tried the image, the exec works well, except one minor issue: it doesn't respond to "tab". Do you have any idea?

erjadi replied:
Whoops you found a bug. I should check how many bytes were actually read from stdin. I was sending 512 bytes every time, mostly zeroes. Apparently that breaks tab functionality.

Fixed it, will push.

robbiezhang commented:
@erjadi appreciate the great work. I tried your image, and it works well. I left some comments in the PR.
For the 1st TODO, I would recommend to add a GetPodFullName function in the interface to let the provider give the right format for the resource name. Can you elaborate the impact of the 2nd TODO?

erjadi commented on vkubelet/apiserver.go:
```diff
@@ -64,3 +68,36 @@ func ApiServerHandler(w http.ResponseWriter, req *http.Request) {
+	idleTimeout := time.Second * 30
+	streamCreationTimeout := time.Second * 30
+
+	remotecommand.ServeExec(w, req, p, fmt.Sprintf("%s-%s", namespace, pod), "", container, []string{command}, streamOpts, idleTimeout, streamCreationTimeout, supportedStreamProtocols)
```
Sounds good, will implement it + provider stubs.

erjadi commented:
Ok I fixed the following:
- Added error handling (closing the streams)
- Tab handling / proper buffer length
- Added GetPodFullName provider function + stubs

btw I noticed that not all the providers are instantiated in provider.go (huawei, azurebatch, mock, vic are missing). Should I add those in this PR or rather open a separate one?

robbiezhang commented:
lgtm
